process_report/process_report.py

```diff
@@ -1,9 +1,108 @@
+
+    old_pi_list = set()
+    try:
+        with open(os.getenv("C0002_OLD_PI")) as f:
+            for pi in f: old_pi_list.add(pi.strip())
+
+    except Exception:
+        print("Old PI file does not exist, or the C0002_OLD_PI env var is not set")
+        sys.exit(1)
+
+    pi_list = dataframe[MANAGER_PI_FIELD].unique()
+
+    for pi in pi_list:
+        if pi != pi: continue   # NaN check
+        if pi in old_pi_list: continue  # Is the PI an old PI?
+
+        pi_projects = dataframe[dataframe[MANAGER_PI_FIELD] == pi]
+        rem_credit = credit_amount
+        for i, row in pi_projects.iterrows():
+            project_cost = row[COST_FIELD]
+            if project_cost >= rem_credit:
+                dataframe.at[i, CREDIT_FIELD] =  rem_credit
+                dataframe.at[i, CREDIT_CODE_FIELD] = credit_code
+                dataframe.at[i, BALANCE_FIELD] = row[COST_FIELD] - rem_credit
+                break
+            else:
+                dataframe.at[i, CREDIT_FIELD] =  project_cost
+                dataframe.at[i, CREDIT_CODE_FIELD] = credit_code
+                dataframe.at[i, BALANCE_FIELD] = 0
+                rem_credit -= project_cost
+    
+    return dataframe
+
+
+applied_credits = [apply_credits_0002]
+
+
+def get_institution_from_pi(pi_uname):
+    
+    institute_map = {
+        "northeastern.edu"      : "Northeastern University",
+        "bu.edu"                : "Boston University",
+        "bentley.edu"           : "Bentley",
+        "uri.edu"               : "University of Rhode Island",
+        "redhat.com"            : "Red Hat",
+        "childrens.harvard.edu" : "Boston Childrens Hospital",
+        "mclean.harvard.edu"    : "McLean Hospital",
+        "meei.harvard.edu"      : "Massachusetts Eye & Ear",
+        "dfci.harvard.edu"      : "Dana-Farber Cancer Institute",
+        "bwh.harvard.edu"       : "Brigham and Women's Hospital",
+        "bidmc.harvard.edu"     : "Beth Israel Deaconess Medical Center",
+        "harvard.edu"           : "Harvard University",
+        "wpi.edu"               : "Worcester Polytechnic Institute",
+        "mit.edu"               : "Massachusetts Institute of Technology",
+        "umass.edu"             : "University of Massachusetts Amherst",
+        "uml.edu"               : "University of Massachusetts Lowell",
+        "codeforboston.org"     : "Code For Boston",
+        "mmsh"                  : "Harvard University",
+        "gstuart"               : "University of Massachusetts Amherst",
+        "rudolph"               : "Boston Childrens Hospital",
+        "robbaron"              : "Boston University",
+        "kmdalton"              : "Harvard University",
+        "mzink"                 : "University of Massachusetts Amherst",
+        "yale.edu"              : "Yale University",
+        "francesco.pontiggia"   : "Harvard University",
+    }
+
+    for name, institute in institute_map.items():
+        if name in pi_uname: return institute
```

Comment:
what's an example of a `pi_uname`? curious how it looks like.

Reply:
It looks like the order of the the dictionary matters in the way this loop works.

For instance, if we have a PI `somebody@dfci.harvard.edu` then `dfci.harvard.edu` must be checked before `harvard.edu` because `name in pi_uname` is true for both cases.  So whoever is modifying the institute mapping file will have to be quite careful to not undo this order. 

I think this would be a faster (no loop) and safer (looking for exact match) way to do this: 

```python3
def get_institution_from_pi(pi_uname, institute_map):
if '@' in pi_uname:
    domain = pi_uname.split('@')[1]
    return institute_map.get(domain, '')

return institute_map.get(pi_uname, '')
```

What do you think?

Reply:
@naved001 After applying your suggestion and running tests on my script again... By looking for exact matches instead of substrings, we might miss cases where the PIs email domain might be a subdomain of an institution (i.e `fas.harvard.edu` is still Harvard University, not an affiliate). This would mean we would have to add more cases for subdomains of the same institution. Do you think it's still worth it to do exact matches.

Reply:
We already have a bunch of subdomains in the institution_map, I think it's okay to update the institution_map with new subdomains when we encounter those.

I thought maybe we could extract the top level domain from the string, but apparently that requires querying all TLDs  which I think is not worth it in our case. See: https://github.com/john-kurkowski/tldextract?tab=readme-ov-file

Reply:
I've applied your suggestion in the [other](https://github.com/CCI-MOC/process_csv_report/pull/12) PR